lib/ClangImporter/ImportDecl.cpp

```diff
@@ -3371,9 +3378,44 @@ namespace {
               decl);
         }
       }
+    }
 
-      return importFunctionDecl(decl, importedName, correctSwiftName,
-                                std::nullopt);
+    /// Emit warnings for unannotated C++ APIs returning FRTs
+    void
+    diagnoseUnannotatedCxxApisReturningFRT(const clang::FunctionDecl *decl) {
+      bool returnsRetainedUnretainedDiagnosticsNeeded = false;
+      if (!isa<clang::CXXDeductionGuideDecl>(decl)) {
+        if (const auto *methodDecl = dyn_cast<clang::CXXMethodDecl>(decl)) {
+          if (canBeRefCountConventionAnnotated(methodDecl)) {
+            // normal c++ member method
+            returnsRetainedUnretainedDiagnosticsNeeded = true;
+          }
+        } else {
+          // global or top-level function
+          returnsRetainedUnretainedDiagnosticsNeeded = true;
+        }
+      }
+
+      if (returnsRetainedUnretainedDiagnosticsNeeded) {
+        HeaderLoc loc(decl->getLocation());
+        Impl.diagnose(loc, diag::no_returns_retained_returns_unretained, decl);
+      }
+    }
+
+    /// Helper routine to exclude things that can't be annotated currently
```

Comment:
I think the comment can be elaborated a little, to make its relationship with the `returns_{,un}retained` attributes more explicit.

```suggestion
    /// Certain C++ methods cannot be annotated with ref-counting conventions
    /// (i.e., "returns_unretained" and "returns_retained"). This helper method
    /// enumerates those cases and returns `false` for them.
```